README.rst
``rutter``:  URL map for Py3k
=============================

.. image:: https://img.shields.io/pypi/v/rutter.svg?style=flat
    :target: https://pypi.python.org/pypi/rutter/
    :alt: Latest Version

.. image:: https://github.com/tseaver/rutter/actions/workflows/ci-tests.yml/badge.svg?branch=master
    :target: https://github.com/tseaver/rutter/actions/workflows/ci-tests.yml

.. image:: https://readthedocs.org/projects/rutter/badge/?version=latest
    :target: https://rutter.readthedocs.org/en/latest/
    :alt: Documentation Status

This project provides a version of the ``URLMap`` class from ``paste.urlmap``,
ported to support Python3.  It also features 100% unit test coverage.

Please see https://rutter.rtfd.io/ for complete documentation


CHANGES.rst
``rutter`` Changelog
====================

1.0 (2023-01-23)
----------------

- No changes from 1.0b1.

1.0b1 (2023-01-18)
------------------

- Add support for Python 3.11.

- Drop support for Python 2.7, 3.5, and 3.6.

0.4 (2022-03-12)
----------------

- Add support for Python 3.9, and 3.10.

- Remove DeprecationWarning from using ``collections.abc``.

- Move tests to GitHub actions.

- Move test suite to py.test.

0.3 (2020-08-10)
----------------

- Add support for Python 3.5, 3.6, 3.7, and 3.8.

- Drop support for Python 2.6, 3.2, 3,3, and 3.4.

- Add support for PyPy3.

0.2 (2013-10-21)
----------------

- Fix generation of ``long-description`` PyPI metadata.

- Add Sphinx documentation.

0.1 (2013-10-16)
----------------

- Initial version.



rutter/tests/test_urlmap.py
import unittest
from webtest import TestApp


class Test__parse_path_expression(unittest.TestCase):

    def _callFUT(self, path):
        from ..urlmap import _parse_path_expression
        return _parse_path_expression(path)

    def test_w_domain_empty(self):
        self.assertRaises(ValueError,
                          self._callFUT, 'domain')

    def test_w_domain_port_empty(self):
        self.assertRaises(ValueError,
                          self._callFUT, 'domain example.com port')

    def test_w_domain_duplicate(self):
        self.assertRaises(ValueError,
                          self._callFUT,
                          'domain example.com domain another.com')

    def test_w_port_only(self):
        self.assertRaises(ValueError,
                          self._callFUT, 'port 80')

    def test_w_port_duplicate(self):
        self.assertRaises(ValueError,
                          self._callFUT,
                          'domain example.com port 80 port 81')

    def test_w_path_duplicate(self):
        self.assertRaises(ValueError,
                          self._callFUT, '/foo /bar')

    def test_wo_domain_or_port(self):
        self.assertEqual(self._callFUT('/foo/bar'), '/foo/bar')

    def test_wo_domain_no_port(self):
        self.assertEqual(self._callFUT('domain example.com /foo/bar'),
                         'http://example.com/foo/bar')

    def test_wo_domain_and_port(self):
        self.assertEqual(self._callFUT('domain example.com port 81 /foo/bar'),
                         'http://example.com:81/foo/bar')

    def test_wo_domain_and_port_no_leading_slash(self):
        self.assertEqual(self._callFUT('domain example.com port 81 foo/bar'),
                         'http://example.com:81/foo/bar')


class Test__normalize_url(unittest.TestCase):

    def _callFUT(self, path, trim=None):
        from ..urlmap import _normalize_url
        if trim is None:
            return _normalize_url(path)
        return _normalize_url(path, trim)

    def test_w_tuple(self):
        domain, path = self._callFUT(('example.com', '/foo'))
        self.assertEqual(domain, 'example.com')
        self.assertEqual(path, '/foo')

    def test_w_list(self):
        domain, path = self._callFUT(['example.com', '/foo'])
        self.assertEqual(domain, 'example.com')
        self.assertEqual(path, '/foo')

    def test_w_string_w_http_domain(self):
        domain, path = self._callFUT('http://example.com/')
        self.assertEqual(domain, 'example.com')
        self.assertEqual(path, '')

    def test_w_string_w_https_domain(self):
        domain, path = self._callFUT('https://example.com/')
        self.assertEqual(domain, 'example.com')
        self.assertEqual(path, '')

    def test_w_string_wo_path(self):
        domain, path = self._callFUT('http://example.com')
        self.assertEqual(domain, 'example.com')
        self.assertEqual(path, '')

    def test_w_empty_string(self):
        domain, path = self._callFUT('')
        self.assertEqual(domain, None)
        self.assertEqual(path, '')

    def test_w_empty_path_w_trim(self):
        domain, path = self._callFUT('/')
        self.assertEqual(domain, None)
        self.assertEqual(path, '')

    def test_w_empty_path_wo_trim(self):
        domain, path = self._callFUT('/', trim=False)
        self.assertEqual(domain, None)
        self.assertEqual(path, '/')

    def test_w_non_slash_path(self):
        self.assertRaises(ValueError, self._callFUT, 'foo')

    def test_w_non_empty_path_w_trim(self):
        domain, path = self._callFUT('/foo/')
        self.assertEqual(domain, None)
        self.assertEqual(path, '/foo')

    def test_w_non_empty_path_wo_trim(self):
        domain, path = self._callFUT('/foo/', trim=False)
        self.assertEqual(domain, None)
        self.assertEqual(path, '/foo/')


class Test__default_not_found_app(unittest.TestCase):

    def _callFUT(self, environ, start_response):
        from ..urlmap import _default_not_found_app
        return _default_not_found_app(environ, start_response)

    def test_wo_mapper(self):
        environ = _makeEnviron()
        _started = []
        def _start_response(status, headers):
            _started.append((status, headers))
        result = list(self._callFUT(environ, _start_response))
        self.assertEqual(len(result), 1)
        lines = result[0].splitlines()
        self.assertEqual(lines[0], b'404 Not Found')
        self.assertEqual(lines[1], b'')
        self.assertEqual(lines[2], b'The resource could not be found.')
        self.assertEqual(lines[3], b'')
        self.assertEqual(lines[4], b" /  " +
                                   b" SCRIPT_NAME: ''" +
                                   b" PATH_INFO: '/'" +
                                   b" HTTP_HOST: 'example.com'  ")

    def test_w_mapper(self):
        _FOO, _BAR = object(), object()
        class _Mapper(object):
            applications = [('/foo', _FOO), ('/bar', _BAR)]
        environ = _makeEnviron()
        environ['paste.urlmap_object'] = _Mapper()
        _started = []
        def _start_response(status, headers):
            _started.append((status, headers))
        result = list(self._callFUT(environ, _start_response))
        self.assertEqual(len(result), 1)
        lines = result[0].splitlines()
        self.assertEqual(lines[0], b'404 Not Found')
        self.assertEqual(lines[1], b'')
        self.assertEqual(lines[2], b'The resource could not be found.')
        self.assertEqual(lines[3], b'')
        self.assertEqual(lines[4], b" / " +
                                   b" defined apps: '/foo',   '/bar'" +
                                   b" SCRIPT_NAME: ''" +
                                   b" PATH_INFO: '/'" +
                                   b" HTTP_HOST: 'example.com'  ")


class URLMapTests(unittest.TestCase):

    def _getTargetClass(self):
        from ..urlmap import URLMap
        return URLMap

    def _makeOne(self, not_found_app=None):
        if not_found_app is None:
            return self._getTargetClass()()
        return self._getTargetClass()(not_found_app)

    def test_ctor_wo_not_found_app(self):
        from ..urlmap import _default_not_found_app
        mapper = self._makeOne()
        self.assertEqual(mapper.applications, [])
        self.assertTrue(mapper.not_found_application is _default_not_found_app)

    def test_ctor_w_not_found_app(self):
        _NOT_FOUND = object()
        mapper = self._makeOne(_NOT_FOUND)
        self.assertEqual(mapper.applications, [])
        self.assertTrue(mapper.not_found_application is _NOT_FOUND)

    def test__sort_apps_empty(self):
        mapper = self._makeOne()
        mapper._sort_apps()
        self.assertEqual(mapper.applications, [])

    def test__sort_apps_non_empty(self):
        _APP1, _APP2, _APP3 = object(), object(), object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo'), _APP1))
        mapper.applications.append(((None, '/foo/bar'), _APP2))
        mapper.applications.append(((None, '/foobar'), _APP3))
        mapper._sort_apps()
        self.assertEqual(mapper.applications,
                         [((None, '/foo/bar'), _APP2),
                          ((None, '/foobar'), _APP3),
                          ((None, '/foo'), _APP1),
                         ])

    def test__sort_apps_non_empty_w_domains(self):
        _APP1, _APP2, _APP3, _APP4 = object(), object(), object(), object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo'), _APP1))
        mapper.applications.append(((None, '/foo/bar'), _APP2))
        mapper.applications.append(((None, '/foobar'), _APP3))
        mapper.applications.append((('example.com', '/foo'), _APP4))
        mapper._sort_apps()
        self.assertEqual(mapper.applications,
                         [(('example.com', '/foo'), _APP4),
                          ((None, '/foo/bar'), _APP2),
                          ((None, '/foobar'), _APP3),
                          ((None, '/foo'), _APP1),
                         ])

    def test___getitem___miss(self):
        mapper = self._makeOne()
        def _test():
            return mapper['/nonesuch']
        self.assertRaises(KeyError, _test)

    def test___getitem___hit(self):
        _APP1, _APP2 = object(), object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo/bar'), _APP2))
        mapper.applications.append(((None, '/foo'), _APP1))
        self.assertTrue(mapper['/foo'] is _APP1)
        self.assertTrue(mapper['/foo/bar'] is _APP2)

    def test___setitem___w_app_None_miss(self):
        mapper = self._makeOne()
        mapper[(None, '/nonesuch')] = None # no raise
        self.assertEqual(mapper.applications, [])

    def test___setitem___w_app_None_hit(self):
        _APP1 = object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo'), _APP1))
        mapper[(None, '/foo')] = None
        self.assertEqual(mapper.applications, [])

    def test___setitem___wo_existing(self):
        _APP1 = object()
        mapper = self._makeOne()
        mapper['/foo'] = _APP1
        self.assertEqual(mapper.applications, [((None, '/foo'), _APP1)])

    def test___setitem___w_existing(self):
        _APP1, _APP2 = object(), object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo'), _APP1))
        mapper['/foo'] = _APP2
        self.assertEqual(mapper.applications, [((None, '/foo'), _APP2)])

    def test___setitem___sorts(self):
        _APP1, _APP2, _APP3, _APP4 = object(), object(), object(), object()
        mapper = self._makeOne()
        mapper['/foo'] = _APP1
        mapper['/foo/bar'] = _APP2
        mapper['/foobar'] = _APP3
        mapper['http://example.com/foo'] = _APP4
        self.assertEqual(mapper.applications,
                         [(('example.com', '/foo'), _APP4),
                          ((None, '/foo/bar'), _APP2),
                          ((None, '/foobar'), _APP3),
                          ((None, '/foo'), _APP1),
                         ])

    def test___delitem___miss(self):
        mapper = self._makeOne()
        def _test():
            del mapper['/nonesuch']
        self.assertRaises(KeyError, _test)

    def test___delitem___hit(self):
        _APP1, _APP2 = object(), object()
        mapper = self._makeOne()
        mapper.applications.append(((None, '/foo/bar'), _APP2))
        mapper.applications.append(((None, '/foo'), _APP1))
        del mapper['/foo']
        self.assertEqual(mapper.applications,
                         [((None, '/foo/bar'), _APP2),
                         ])

    def test_keys_empty(self):
        mapper = self._makeOne()
        self.assertEqual(mapper.keys(), [])

    def test_keys_non_empty(self):
        _APP1, _APP2, _APP3, _APP4 = object(), object(), object(), object()
        mapper = self._makeOne()
        mapper['/foo'] = _APP1
        mapper['/foo/bar'] = _APP2
        mapper['/foobar'] = _APP3
        mapper['http://example.com/foo'] = _APP4
        self.assertEqual(mapper.keys(),
                         [('example.com', '/foo'),
                          (None, '/foo/bar'),
                          (None, '/foobar'),
                          (None, '/foo'),
                         ])

    def test___iter___empty(self):
        mapper = self._makeOne()
        self.assertEqual(list(mapper), [])

    def test___iter___non_empty(self):
        _APP1, _APP2, _APP3, _APP4 = object(), object(), object(), object()
        mapper = self._makeOne()
        mapper['/foo'] = _APP1
        mapper['/foo/bar'] = _APP2
        mapper['/foobar'] = _APP3
        mapper['http://example.com/foo'] = _APP4
        self.assertEqual(list(mapper),
                         [('example.com', '/foo'),
                          (None, '/foo/bar'),
                          (None, '/foobar'),
                          (None, '/foo'),
                         ])

    def test___len___empty(self):
        mapper = self._makeOne()
        self.assertEqual(len(mapper), 0)

    def test___len___non_empty(self):
        _APP1, _APP2, _APP3, _APP4 = object(), object(), object(), object()
        mapper = self._makeOne()
        mapper['/foo'] = _APP1
        mapper['/foo/bar'] = _APP2
        mapper['/foobar'] = _APP3
        mapper['http://example.com/foo'] = _APP4
        self.assertEqual(len(mapper), 4)

    def test___call___w_empty(self):
        not_found = DummyApp()
        environ = _makeEnviron()
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        result = mapper(environ, _start_response)
        self.assertTrue(result is not_found)
        self.assertTrue(environ['paste.urlmap_object'] is mapper)
        self.assertTrue(result.environ is environ)
        self.assertTrue(result.start_response is _start_response)

    def test___call___w_HTTP_HOST_and_SERVER_NAME(self):
        not_found = DummyApp()
        never_called = DummyApp()
        environ = _makeEnviron(SERVER_NAME='otherdomain.com')
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://otherdomain.com/'] = never_called
        result = mapper(environ, _start_response)
        self.assertTrue(result is not_found)
        self.assertTrue(never_called.environ is None)

    def test___call___w_only_SERVER_NAME(self):
        not_found = DummyApp()
        called = DummyApp()
        environ = _makeEnviron(SERVER_NAME='Example.com')
        del environ['HTTP_HOST']
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com/'] = called
        result = mapper(environ, _start_response)
        self.assertTrue(result is called)
        self.assertTrue(not_found.environ is None)

    def test___call___w_port_in_HTTP_HOST(self):
        not_found = DummyApp()
        never_called = DummyApp()
        environ = _makeEnviron()
        environ['HTTP_HOST'] = 'http://example.com:8080'
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com:8888'] = never_called
        result = mapper(environ, _start_response)
        self.assertTrue(result is not_found)
        self.assertTrue(never_called.environ is None)

    def test___call___wo_port_in_HTTP_HOST_miss(self):
        not_found = DummyApp()
        never_called = DummyApp()
        environ = _makeEnviron()
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com:8888'] = never_called
        result = mapper(environ, _start_response)
        self.assertTrue(result is not_found)
        self.assertTrue(never_called.environ is None)

    def test___call___wo_port_in_HTTP_HOST_https_scheme(self):
        not_found = DummyApp()
        never_called = DummyApp()
        https = DummyApp()
        environ = _makeEnviron()
        environ['wsgi.url_scheme'] = 'https'
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com:80'] = never_called
        mapper['http://example.com:443'] = https
        result = mapper(environ, _start_response)
        self.assertTrue(result is https)
        self.assertTrue(never_called.environ is None)

    def test___call___wo_port_in_HTTP_HOST_hit(self):
        not_found = DummyApp()
        shorter = DummyApp()
        longer = DummyApp()
        environ = _makeEnviron(PATH_INFO='/foobar')
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com:80/foo'] = shorter
        mapper['http://example.com:80/foobar'] = longer
        result = mapper(environ, _start_response)
        self.assertTrue(result is longer)
        self.assertTrue(shorter.environ is None)
        self.assertEqual(environ['SCRIPT_NAME'], '/foobar')
        self.assertEqual(environ['PATH_INFO'], '')

    def test___call___wo_port_in_HTTP_HOST_hit_w_subpath(self):
        not_found = DummyApp()
        shorter = DummyApp()
        longer = DummyApp()
        environ = _makeEnviron(PATH_INFO='/foobar/baz')
        def _start_response(status, headers): pass
        mapper = self._makeOne(not_found)
        mapper['http://example.com:80/foo'] = shorter
        mapper['http://example.com:80/foobar'] = longer
        result = mapper(environ, _start_response)
        self.assertTrue(result is longer)
        self.assertTrue(shorter.environ is None)
        self.assertEqual(environ['SCRIPT_NAME'], '/foobar')
        self.assertEqual(environ['PATH_INFO'], '/baz')


class Test_urlmap_factory(unittest.TestCase):

    def _callFUT(self, loader, global_conf, **local_conf):
        from ..urlmap import urlmap_factory
        return urlmap_factory(loader, global_conf, **local_conf)

    def test_empty_wo_notfound_app(self):
        from ..urlmap import _default_not_found_app
        mapper = self._callFUT(object(), {})
        self.assertEqual(len(mapper), 0)
        self.assertTrue(mapper.not_found_application is _default_not_found_app)

    def test_empty_w_global_notfound_app(self):
        not_found = DummyApp()
        loader = DummyLoader(xxx=not_found)
        gconf = {'not_found_app': 'xxx'}
        before = gconf.copy()
        mapper = self._callFUT(loader, gconf)
        self.assertEqual(len(mapper), 0)
        self.assertTrue(mapper.not_found_application is not_found)
        self.assertEqual(gconf, before)

    def test_empty_w_local_notfound_app(self):
        not_found = DummyApp()
        loader = DummyLoader(yyy=not_found)
        gconf = {'not_found_app': 'xxx'}
        before = gconf.copy()
        mapper = self._callFUT(loader, gconf, not_found_app='yyy')
        self.assertEqual(len(mapper), 0)
        self.assertTrue(mapper.not_found_application is not_found)
        self.assertEqual(gconf, before)

    def test_nonempty(self):
        not_found = DummyApp()
        _APP1, _APP2, _APP3 = DummyApp(), DummyApp(), DummyApp()
        loader = DummyLoader(www=not_found, xxx=_APP1, yyy=_APP2, zzz=_APP3)
        umap = {'/foo': 'xxx', '/foobar': 'yyy', '/foo/bar': 'zzz'}
        mapper = self._callFUT(loader, {}, not_found_app='www', **umap)
        self.assertEqual(len(mapper), 3)
        self.assertTrue(mapper.not_found_application is not_found)
        self.assertTrue(mapper['/foo'] is _APP1)
        self.assertTrue(mapper['/foobar'] is _APP2)
        self.assertTrue(mapper['/foo/bar'] is _APP3)


class DummyApp(object):

    environ = start_response = None

    def __call__(self, environ, start_response):
        self.environ = environ
        self.start_response = start_response
        return self

class DummyLoader(dict):

    def get_app(self, spec, global_conf):
        return self[spec]

def _makeEnviron(**kw):
    environ = {
        'HTTP_HOST': 'example.com',
        'REQUEST_METHOD': 'GET',
        'SCRIPT_NAME': '',
        'PATH_INFO': '/',
        'wsgi.url_scheme': 'http',
    }
    environ.update(kw)
    return environ


class Functests(unittest.TestCase):

    def _makeOne(self):
        from ..urlmap import URLMap

        mapper = URLMap()
        return mapper, TestApp(mapper)

    def test_map(self):

        def _make_app(response_text):
            def app(environ, start_response):
                headers = [('Content-type', 'text/html')]
                start_response('200 OK', headers)
                return [(response_text % environ).encode('utf8')]
            return app

        mapper, app = self._makeOne()
        text = '%s script_name="%%(SCRIPT_NAME)s" path_info="%%(PATH_INFO)s"'
        mapper[''] = _make_app(text % 'root')
        mapper['/foo'] = _make_app(text % 'foo-only')
        mapper['/foo/bar'] = _make_app(text % 'foo:bar')
        mapper['/f'] = _make_app(text % 'f-only')
        res = app.get('/')
        res.mustcontain('root')
        res.mustcontain('script_name=""')
        res.mustcontain('path_info="/"')
        res = app.get('/blah')
        res.mustcontain('root')
        res.mustcontain('script_name=""')
        res.mustcontain('path_info="/blah"')
        res = app.get('/foo/and/more')
        res.mustcontain('script_name="/foo"')
        res.mustcontain('path_info="/and/more"')
        res.mustcontain('foo-only')
        res = app.get('/foo/bar/baz')
        res.mustcontain('foo:bar')
        res.mustcontain('script_name="/foo/bar"')
        res.mustcontain('path_info="/baz"')
        res = app.get('/fffzzz')
        res.mustcontain('root')
        res.mustcontain('path_info="/fffzzz"')
        res = app.get('/f/z/y')
        res.mustcontain('script_name="/f"')
        res.mustcontain('path_info="/z/y"')
        res.mustcontain('f-only')

    def test_404(self):
        mapper, app = self._makeOne()
        res = app.get("/-->%0D<script>alert('xss')</script>", status=404)
        self.assertTrue(b'--><script' not in res.body)
        res = app.get("/--%01><script>", status=404)
        self.assertTrue(b'--\x01><script>' not in res.body)


rutter/urlmap.py
""" Map URL prefixes to WSGI applications.  See ``URLMap``

Forked from ``paste.urllib``.
"""
try:
    from collections.abc import MutableMapping
except ImportError:  # pragma: NO COVER Python2
    from collections import MutableMapping
try:
    from html import escape
except ImportError:  # pragma: NO COVER Python2
    from cgi import escape
import re

from webob.exc import HTTPNotFound

def _parse_path_expression(path):
    """ Parse a path expression for a path alone.
    
    E.g., for 'domain foobar.com port 20 /', return '/'; for '/foobar'
    return '/foobar'
    
    Return as an address that URLMap likes.
    """
    parts = path.split()
    domain = port = path = None
    while parts:
        if parts[0] == 'domain':
            parts.pop(0)
            if not parts:
                raise ValueError("'domain' must be followed with a domain name")
            if domain:
                raise ValueError("'domain' given twice")
            domain = parts.pop(0)
        elif parts[0] == 'port':
            parts.pop(0)
            if not parts:
                raise ValueError("'port' must be followed with a port number")
            if port:
                raise ValueError("'port' given twice")
            port = parts.pop(0)
        else:
            if path:
                raise ValueError("more than one path given (have %r, got %r)"
                                 % (path, parts[0]))
            path = parts.pop(0)
    s = ''
    if domain:
        s = 'http://%s' % domain
    if port:
        if not domain:
            raise ValueError("If you give a port, you must also give a domain")
        s += ':' + port
    if path:
        if s and not path.startswith('/'):
            s += '/'
        s += path
    return s


_NORM_URL_RE = re.compile('//+')
_DOMAIN_URL_RE = re.compile('^(http|https)://')

def _default_not_found_app(environ, start_response):
    mapper = environ.get('paste.urlmap_object')
    if mapper:
        matches = [p for p, a in mapper.applications]
        extra = 'defined apps: %s' % (
            ',\n  '.join(map(repr, matches)))
    else:
        extra = ''
    extra += '\nSCRIPT_NAME: %r' % environ.get('SCRIPT_NAME')
    extra += '\nPATH_INFO: %r' % environ.get('PATH_INFO')
    extra += '\nHTTP_HOST: %r' % environ.get('HTTP_HOST')
    # XXX WebOb?
    exc = HTTPNotFound(environ['PATH_INFO'], comment=escape(extra, quote=False))
    return exc(environ, start_response)

def _normalize_url(url, trim=True):
    """Return ``(domain, path)`` tuple for ``url``.

    If ``trim`` is True, remove any trailing slash from ``path``.
    """
    if isinstance(url, (list, tuple)):
        domain = url[0]
        url = _normalize_url(url[1])[1]
        return domain, url
    if not (not url or url.startswith('/')
            or _DOMAIN_URL_RE.search(url)):
        raise ValueError(
            "URL fragments must start with / or http:// (you gave %r)"
            % url)
    match = _DOMAIN_URL_RE.search(url)
    if match:
        url = url[match.end():]
        if '/' in url:
            domain, url = url.split('/', 1)
            url = '/' + url
        else:
            domain, url = url, ''
    else:
        domain = None
    url = _NORM_URL_RE.sub('/', url)
    if trim:
        url = url.rstrip('/')
    return domain, url


class URLMap(MutableMapping):
    """Dispatch to one of several applications based on the URL.

    The dictionary keys are URLs to match (like
    ``PATH_INFO.startswith(url)``), and the values are applications to
    dispatch to.  URLs are matched most-specific-first, i.e., longest
    URL first.  The ``SCRIPT_NAME`` and ``PATH_INFO`` environmental
    variables are adjusted to indicate the new context.

    URLs can also include domains, like ``http://blah.com/foo``, or as
    tuples ``('blah.com', '/foo')``.  This will match domain names; without
    the ``http://domain`` or with a domain of ``None`` any domain will be
    matched (so long as no other explicit domain matches).
    """
    def __init__(self, not_found_app=_default_not_found_app):
        self.applications = []
        self.not_found_application = not_found_app

    def _sort_apps(self):
        """Sort applications, longest URLs first.

        Apps w/o domains sort *last*.
        """
        def key(app_desc):
            dom_url, app = app_desc
            domain, url = dom_url
            return domain or '\xff', -len(url)
        self.applications = sorted(self.applications, key=key)

    def __getitem__(self, url):
        dom_url = _normalize_url(url)
        for app_url, app in self.applications:
            if app_url == dom_url:
                return app
        raise KeyError(
            "No application with the url %r (domain: %r; existing: %s)"
            % (url[1], url[0] or '*', self.applications))

    def __setitem__(self, url, app):
        if app is None:
            try:
                del self[url]
            except KeyError:
                pass
            return
        dom_url = _normalize_url(url)
        if dom_url in self:
            del self[dom_url]
        self.applications.append((dom_url, app))
        self._sort_apps()

    def __delitem__(self, url):
        url = _normalize_url(url)
        for app_url, app in self.applications:
            if app_url == url:
                self.applications.remove((app_url, app))
                break
        else:
            raise KeyError(
                "No application with the url %r" % (url,))

    def keys(self):
        return [app_url for app_url, app in self.applications]

    def __iter__(self):
        return (x[0] for x in self.applications)

    def __len__(self):
        return len(self.applications)

    def __call__(self, environ, start_response):
        host = environ.get('HTTP_HOST', environ.get('SERVER_NAME')).lower()
        if ':' in host:
            host, port = host.split(':', 1)
        else:
            if environ['wsgi.url_scheme'] == 'http':
                port = '80'
            else:
                port = '443'
        hostport = host + ':' + port
        path_info = environ.get('PATH_INFO')
        path_info = _normalize_url(path_info, False)[1]
        for dom_url, app in self.applications:
            domain, app_url = dom_url
            if domain and domain != host and domain != hostport:
                continue
            if (path_info == app_url
                or path_info.startswith(app_url + '/')):
                environ['SCRIPT_NAME'] += app_url
                environ['PATH_INFO'] = path_info[len(app_url):]
                return app(environ, start_response)
        environ['paste.urlmap_object'] = self
        return self.not_found_application(environ, start_response)

def urlmap_factory(loader, global_conf, **local_conf):
    if 'not_found_app' in local_conf:
        not_found_app = local_conf.pop('not_found_app')
    else:
        not_found_app = global_conf.get('not_found_app')
    if not_found_app:
        not_found_app = loader.get_app(not_found_app, global_conf=global_conf)
    if not_found_app is not None:
        urlmap = URLMap(not_found_app=not_found_app)
    else:
        urlmap = URLMap()
    for path, app_name in local_conf.items():
        path = _parse_path_expression(path)
        app = loader.get_app(app_name, global_conf=global_conf)
        urlmap[path] = app
    return urlmap
